File: ISVImageScrollView/Classes/ISVImageScrollView.m
//
//  ISVImageScrollView.m
//  ISVImageScrollView
//
//  Created by Yurii Kupratsevich on 10/23/17.
//

#import "ISVImageScrollView.h"

static void *ScrollViewBoundsChangeNotificationContext = &ScrollViewBoundsChangeNotificationContext;

@interface ISVImageScrollView ()

@property (nonatomic, readonly) CGFloat imageAspectRatio;
@property (nonatomic) CGRect initialImageFrame;
@property (strong, nonatomic, readonly) UITapGestureRecognizer *tap;

@end

@implementation ISVImageScrollView

@synthesize tap = _tap;

#pragma mark - Tap to Zoom

-(UITapGestureRecognizer *)tap {
    if (_tap == nil) {
        _tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapToZoom:)];
        _tap.numberOfTapsRequired = 2;
    }
    return _tap;
}

- (void)tapToZoom:(UIGestureRecognizer *)gestureRecognizer {
    if (self.zoomScale > self.minimumZoomScale) {
        [self setZoomScale:self.minimumZoomScale animated:YES];
    } else {
        CGPoint tapLocation = [gestureRecognizer locationInView:self.imageView];
        CGFloat zoomRectWidth = self.imageView.frame.size.width / self.maximumZoomScale;
        CGFloat zoomRectHeight = self.imageView.frame.size.height / self.maximumZoomScale;
        CGFloat zoomRectX = tapLocation.x - zoomRectWidth * 0.5;
        CGFloat zoomRectY = tapLocation.y - zoomRectHeight * 0.5;
        CGRect zoomRect = CGRectMake(zoomRectX, zoomRectY, zoomRectWidth, zoomRectHeight);
        [self zoomToRect:zoomRect animated:YES];
    }
}


#pragma mark - Private Methods

- (void)configure {
    self.showsVerticalScrollIndicator = NO;
    self.showsHorizontalScrollIndicator = NO;
    [self startObservingBoundsChange];
}

- (void)setImageView:(UIImageView *)imageView {
    if (_imageView.superview == self) {
        [_imageView removeGestureRecognizer:self.tap];
        [_imageView removeFromSuperview];
    }
    if (imageView) {
        _imageView = imageView;
        _initialImageFrame = CGRectNull;
        _imageView.userInteractionEnabled = YES;
        [_imageView addGestureRecognizer:self.tap];
        [self addSubview:imageView];
    }
}

- (CGFloat)imageAspectRatio {
    if (self.imageView.image) {
        return self.imageView.image.size.width / self.imageView.image.size.height;
    }
    return 1;
}

- (CGSize)rectSizeForAspectRatio:(CGFloat)ratio thatFitsSize:(CGSize)size {
    CGFloat containerWidth = size.width;
    CGFloat containerHeight = size.height;
    CGFloat resultWidth = 0;
    CGFloat resultHeight = 0;
    
    if (containerWidth / containerHeight >= ratio) {
        resultHeight = containerHeight;
        resultWidth = containerHeight * ratio;
    }
    else {
        resultWidth = containerWidth;
        resultHeight = containerWidth / ratio;
    }
    
    return CGSizeMake(resultWidth, resultHeight);
}

- (void)scaleImageForScrollViewTransitionFromBounds:(CGRect)oldBounds toBounds:(CGRect)newBounds {
    CGPoint oldContentOffset = CGPointMake(oldBounds.origin.x, oldBounds.origin.y);
    CGSize oldSize = oldBounds.size;
    CGSize newSize = newBounds.size;
    
    CGSize containedImageSizeOld = [self rectSizeForAspectRatio:self.imageAspectRatio
                                                   thatFitsSize:oldSize];
    
    CGSize containedImageSizeNew = [self rectSizeForAspectRatio:self.imageAspectRatio
                                                   thatFitsSize:newSize];
    
    CGFloat orientationRatio = containedImageSizeNew.height / containedImageSizeOld.height;
    
    CGAffineTransform t = CGAffineTransformMakeScale(orientationRatio, orientationRatio);
    self.imageView.frame = CGRectApplyAffineTransform(self.imageView.frame, t);
    
    self.contentSize = self.imageView.frame.size;
    
    CGFloat xOffset = (oldContentOffset.x + oldSize.width * 0.5) * orientationRatio - newSize.width * 0.5;
    CGFloat yOffset = (oldContentOffset.y + oldSize.height * 0.5) * orientationRatio - newSize.height * 0.5;
    
    xOffset -= MAX(xOffset + newSize.width - self.contentSize.width, 0);
    yOffset -= MAX(yOffset + newSize.height - self.contentSize.height, 0);
    xOffset += MAX(-xOffset, 0);
    yOffset += MAX(-yOffset, 0);
    
    self.contentOffset = CGPointMake(xOffset, yOffset);
}

- (void)setupInitialImageFrame {
    if (self.imageView.image && CGRectEqualToRect(self.initialImageFrame, CGRectNull)) {
        CGSize imageViewSize = [self rectSizeForAspectRatio:self.imageAspectRatio
                                               thatFitsSize:self.bounds.size];
        self.initialImageFrame = CGRectMake(0, 0, imageViewSize.width, imageViewSize.height);
        self.imageView.frame = self.initialImageFrame;
        self.contentSize = self.initialImageFrame.size;
    }
}

#pragma mark - KVO

- (void)startObservingBoundsChange {
    [self addObserver:self
           forKeyPath:@"bounds"
              options:NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew
              context:ScrollViewBoundsChangeNotificationContext];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if (context == ScrollViewBoundsChangeNotificationContext) {
        CGRect oldRect = ((NSValue *)change[NSKeyValueChangeOldKey]).CGRectValue;
        CGRect newRect = ((NSValue *)change[NSKeyValueChangeNewKey]).CGRectValue;
        if (! CGSizeEqualToSize(oldRect.size, newRect.size)) {
            [self scaleImageForScrollViewTransitionFromBounds:oldRect toBounds:newRect];
        }
    }
}

- (void)dealloc {
    [self removeObserver:self forKeyPath:@"bounds"];
}

#pragma mark - UIScrollView

- (void)setContentOffset:(CGPoint)contentOffset
{
    const CGSize contentSize = self.contentSize;
    const CGSize scrollViewSize = self.bounds.size;
    
    if (contentSize.width < scrollViewSize.width)
    {
        contentOffset.x = - (scrollViewSize.width - contentSize.width) * 0.5;
    }
    
    if (contentSize.height < scrollViewSize.height)
    {
        contentOffset.y = - (scrollViewSize.height - contentSize.height) * 0.5;
    }
    
    [super setContentOffset:contentOffset];
}

#pragma mark - UIView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self configure];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self configure];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    [self setupInitialImageFrame];
}

@end

File: Sources/ISVImageScrollView/ISVImageScrollView.swift
//
//  ISVImageScrollView.swift
//  ISVImageScrollView_Example
//
//  Created by Yurii Kupratsevych on 17.06.2020.
//  Copyright © 2020 kupratsevich@gmail.com. All rights reserved.
//

import UIKit

public class ISVImageScrollView: UIScrollView, UIGestureRecognizerDelegate {
  
  // MARK: - Public
  
  public var imageView: UIImageView? {
    didSet {
      oldValue?.removeGestureRecognizer(self.tap)
      oldValue?.removeFromSuperview()
      if let imageView = self.imageView {
        self.initialImageFrame = .null
        imageView.isUserInteractionEnabled = true
        imageView.addGestureRecognizer(self.tap)
        self.addSubview(imageView)
      }
    }
  }
  
  // MARK: - Initialization
  
  override init(frame: CGRect) {
    super.init(frame: frame)
    self.configure()
  }
  
  required init?(coder: NSCoder) {
    super.init(coder: coder)
    self.configure()
  }
  
  deinit {
    self.stopObservingBoundsChange()
  }
  
  // MARK: - UIScrollView
  
  public override func layoutSubviews() {
    super.layoutSubviews()
    self.setupInitialImageFrame()
  }
  
  public override var contentOffset: CGPoint {
    didSet {
      let contentSize = self.contentSize
      let scrollViewSize = self.bounds.size
      var newContentOffset = contentOffset
      
      if contentSize.width < scrollViewSize.width {
        newContentOffset.x = (contentSize.width - scrollViewSize.width) * 0.5
      }
      
      if contentSize.height < scrollViewSize.height {
        newContentOffset.y = (contentSize.height - scrollViewSize.height) * 0.5
      }
      
      super.contentOffset = newContentOffset
    }
  }
  
  // MARK: - UIGestureRecognizerDelegate
  
  public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
    return otherGestureRecognizer === self.panGestureRecognizer
  }
  
  // MARK: - Private: Tap to Zoom
  
  private lazy var tap: UITapGestureRecognizer = {
    let tap = UITapGestureRecognizer(target: self, action: #selector(tapToZoom(_:)))
    tap.numberOfTapsRequired = 2
    tap.delegate = self
    return tap
  }()
  
  @IBAction private func tapToZoom(_ sender: UIGestureRecognizer) {
    guard sender.state == .ended else { return }
    if self.zoomScale > self.minimumZoomScale {
      self.setZoomScale(self.minimumZoomScale, animated: true)
    } else {
      guard let imageView = self.imageView else { return }
      let tapLocation = sender.location(in: imageView)
      let zoomRectWidth = imageView.frame.size.width / self.maximumZoomScale;
      let zoomRectHeight = imageView.frame.size.height / self.maximumZoomScale;
      let zoomRectX = tapLocation.x - zoomRectWidth * 0.5;
      let zoomRectY = tapLocation.y - zoomRectHeight * 0.5;
      let zoomRect = CGRect(
        x: zoomRectX,
        y: zoomRectY,
        width: zoomRectWidth,
        height: zoomRectHeight)
      self.zoom(to: zoomRect, animated: true)
    }
  }

  // MARK: - Private: Geometry
  
  private var initialImageFrame: CGRect = .null
  
  private var imageAspectRatio: CGFloat {
    guard let image = self.imageView?.image else { return 1 }
    return image.size.width / image.size.height
  }
  
  private func configure() {
    self.showsVerticalScrollIndicator = false
    self.showsHorizontalScrollIndicator = false
    self.startObservingBoundsChange()
  }
  
  private func rectSize(for aspectRatio: CGFloat, thatFits size: CGSize) -> CGSize {
    let containerWidth = size.width
    let containerHeight = size.height
    var resultWidth: CGFloat = 0
    var resultHeight: CGFloat = 0
    
    if aspectRatio <= 0 || containerHeight <= 0 {
      return size
    }
    
    if containerWidth / containerHeight >= aspectRatio {
      resultHeight = containerHeight
      resultWidth = containerHeight * aspectRatio
    } else {
      resultWidth = containerWidth
      resultHeight = containerWidth / aspectRatio
    }
    
    return CGSize(width: resultWidth, height: resultHeight)
  }
  
  private func scaleImageForTransition(from oldBounds: CGRect, to newBounds: CGRect) {
    guard let imageView = self.imageView else { return}
    
    let oldContentOffset = CGPoint(x: oldBounds.origin.x, y: oldBounds.origin.y)
    let oldSize = oldBounds.size
    let newSize = newBounds.size
    var containedImageSizeOld = self.rectSize(for: self.imageAspectRatio, thatFits: oldSize)
    let containedImageSizeNew = self.rectSize(for: self.imageAspectRatio, thatFits: newSize)
    
    if containedImageSizeOld.height <= 0 {
      containedImageSizeOld = containedImageSizeNew
    }
    
    let orientationRatio = containedImageSizeNew.height / containedImageSizeOld.height
    let transform = CGAffineTransform(scaleX: orientationRatio, y: orientationRatio)
    self.imageView?.frame = imageView.frame.applying(transform)
    self.contentSize = imageView.frame.size;
    
    var xOffset = (oldContentOffset.x + oldSize.width * 0.5) * orientationRatio - newSize.width * 0.5
    var yOffset = (oldContentOffset.y + oldSize.height * 0.5) * orientationRatio - newSize.height * 0.5
    
    xOffset -= max(xOffset + newSize.width - self.contentSize.width, 0)
    yOffset -= max(yOffset + newSize.height - self.contentSize.height, 0)
    xOffset -= min(xOffset, 0)
    yOffset -= min(yOffset, 0)
    
    self.contentOffset = CGPoint(x: xOffset, y: yOffset)
  }
  
  private func setupInitialImageFrame() {
    guard self.imageView != nil, self.initialImageFrame == .null else { return }
    let imageViewSize = self.rectSize(for: self.imageAspectRatio, thatFits: self.bounds.size)
    self.initialImageFrame = CGRect(x: 0, y: 0, width: imageViewSize.width, height: imageViewSize.height)
    self.imageView?.frame = self.initialImageFrame
    self.contentSize = self.initialImageFrame.size
  }
  
  // MARK: - Private: KVO
  
  private var boundsObserver: NSKeyValueObservation?
  
  private func startObservingBoundsChange() {
    self.boundsObserver = self.observe(
      \.self.bounds,
      options: [.old, .new],
      changeHandler: { [weak self] (object, change) in
        if let oldRect = change.oldValue,
          let newRect = change.newValue,
          oldRect.size != newRect.size {
          self?.scaleImageForTransition(from: oldRect, to: newRect)
        }
    })
  }
  
  private func stopObservingBoundsChange() {
    self.boundsObserver?.invalidate()
    self.boundsObserver = nil
  }
}
